control-plane/csi-driver/src/bin/controller/controller.rs

```diff
@@ -74,6 +78,66 @@ fn parse_protocol(proto: Option<&String>) -> Result<VolumeShareProtocol, Status>
     }
 }
 
+/// Get app node where volume frontend is situated.
+fn get_volume_app_node(volume: &Volume) -> Option<String> {
+    match volume.spec.target.as_ref() {
+        // todo: Might need to deal with multiple app nodes later?
+        Some(target) => target
+            .frontend_nodes
+            .as_ref()
+            .map(|nodes| nodes[0].name.clone()),
+        None => None,
+    }
+}
+
+async fn issue_fs_freeze(endpoint: String, volume_id: String) -> Result<(), Status> {
+    trace!(
+        "Issuing fs freeze for volume {}, endpoint {}",
+        volume_id,
+        endpoint
+    );
+    let mut client = NodePluginClient::connect(format!("http://{endpoint}"))
+        .await
+        .map_err(|error| Status::failed_precondition(error.to_string()))?;
+    match client
+        .freeze_fs(Request::new(FreezeFsRequest {
+            volume_id: volume_id.clone(),
+        }))
+        .await
+    {
+        Ok(_) => Ok(()),
+        Err(status) if status.code() == tonic::Code::Unimplemented => {
+            trace!("fs unfreeze not supported for raw block volume: {volume_id}");
```

Comment:
If we add instrument above, then volume_id is automatically added.
```suggestion
            trace!("fs unfreeze not supported for raw block volume");
```
I wonder if not supported or something like that is a better error code here

Reply:
```
    /// Operation is not implemented or not supported.
    Unimplemented = 12,
```
    
  unimplement itself is for not supported as per docs

Reply:
The problem is that unimplemented in CSI is usually for the operation itself, not for specifics, so this could be seen if you're running new csi-controller and old csi-node for example.

Reply:
Wonder if we should be doing a sync for the block volumes as well (not related to this PR)

Reply:
Yeah, any tonic status code that suits here?

Reply:
Yeah we can add sync for block volumes. I will take that up as a separate feature request as that's not a part of this.

Reply:
Probably invalid argument?